### PR TITLE
Add `log_text`

### PR DESCRIPTION
**Pull request by harupy**
## What changes are proposed in this pull request?

Add a new tracking API `log_text` which logs a text as an artifact.

```
mlflow.log_text("a", "b.txt")
```

## How is this patch tested?

Unit test

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add a new fluent API `mlflow.log_text` which logs in-memory text as an artifact.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/client.py [hunk 1]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
+            run = client.create_run(experiment_id="0")
+
+            # Log a text in the artifact root
+            client.log_text(run.info.run_id, "text1", "file1.txt")
+
+            # Log a text in a subdirectory
+            client.log_text(run.info.run_id, "text2", "dir/file2.txt")
+
+            # Log an HTML text
+            client.log_text(run.info.run_id, "<h1>header</h1>", "index.html")
```
Comment (harupy) [inline mlflow/tracking/client.py:955]: Added this example to indicate that this API accepts any text file formats (e.g. .html, .csv, .py, .sh)

File: mlflow/tracking/client.py [hunk 2]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
         self._tracking_client.log_artifacts(run_id, local_dir, artifact_path)
 
+    @contextlib.contextmanager
+    def _log_artifact_helper(self, run_id, artifact_file):
+        """
+        Yields a temporary path to store a file, and then calls `log_artifact` against that path.
+        """
+        norm_path = posixpath.normpath(artifact_file)
+        filename = posixpath.basename(norm_path)
+        artifact_path = posixpath.dirname(norm_path)
```
Comment (harupy) [inline mlflow/tracking/client.py:924]: Used `normpath` here to handle a relative path like `"./a.txt"` or `"a/../b.txt"`.

File: mlflow/tracking/client.py [hunk 3]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
+            yield tmp_path
+            self.log_artifact(run_id, tmp_path, artifact_path)
+
+    def log_text(self, run_id, text, artifact_file):
+        """
+        Log a text as an artifact.
+
+        :param run_id: String ID of the run.
+        :param text: Text to log.
+        :param artifact_file: The run-relative artifact file path to which the text is saved.
```
Comment (harupy) [inline mlflow/tracking/client.py:938]: Should we allow a windows path (e.g. `"foo\\bar.txt"`) as `artifact_file `?

Comment (harupy) [reply mlflow/tracking/client.py:938]: `mlflow.log_artifact` doesn't accept a windows-style path:

```python
import mlflow


with mlflow.start_run() as run:
    mlflow.log_artifact("README.rst", "foo\\bar.txt")

    client = mlflow.tracking.MlflowClient()
    files = [a.path for a in client.list_artifacts(run.info.run_id)]
    print(files)
```

output:

```
['foo\\bar.txt']
```

File: mlflow/tracking/fluent.py [hunk 4]
```diff
@@ -557,6 +557,32 @@ def log_artifacts(local_dir, artifact_path=None):
     MlflowClient().log_artifacts(run_id, local_dir, artifact_path)
 
 
+def log_text(text, artifact_file):
```
Comment (harupy) [inline mlflow/tracking/fluent.py:560]: Should we mark this API as experimental?

File: mlflow/tracking/client.py (hunk 3 above)
Comment (smurching) [reply mlflow/tracking/client.py:938]: I don't think we need to support windows-style paths (technically the `artifact_file` is like a suffix to append to the run's artifact URI), but we should just make it clear in the `artifact_file` docstring that the specified path is expected to be in posixpath format

File: mlflow/tracking/client.py [hunk 5]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
+    def _log_artifact_helper(self, run_id, artifact_file):
+        """
+        Yields a temporary path to store a file, and then calls `log_artifact` against that path.
+        """
+        norm_path = posixpath.normpath(artifact_file)
+        filename = posixpath.basename(norm_path)
+        artifact_path = posixpath.dirname(norm_path)
+        artifact_path = None if artifact_path == "" else artifact_path
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
```
Comment (smurching) [inline mlflow/tracking/client.py:927]: Seems like we should clean up this directory by calling `cleanup` after logging the artifact? https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

File: mlflow/tracking/client.py (hunk 3 above)
Comment (harupy) [reply mlflow/tracking/client.py:938]: Makes sense. I'll update the `artifact_file` description.

File: mlflow/tracking/fluent.py [hunk 6]
```diff
@@ -557,6 +557,32 @@ def log_artifacts(local_dir, artifact_path=None):
...
+    """
+    Log a text as an artifact.
+
+    :param text: Text to log.
```
Comment (smurching) [inline mlflow/tracking/fluent.py:564]:
```suggestion
    :param text: String text to log.
```

File: mlflow/tracking/fluent.py [hunk 7]
```diff
@@ -557,6 +557,32 @@ def log_artifacts(local_dir, artifact_path=None):
     MlflowClient().log_artifacts(run_id, local_dir, artifact_path)
 
 
+def log_text(text, artifact_file):
+    """
+    Log a text as an artifact.
```
Comment (smurching) [inline mlflow/tracking/fluent.py:562]:
```suggestion
    Log text as an artifact.
```

File: mlflow/tracking/fluent.py (hunk 4 above)
Comment (smurching) [reply mlflow/tracking/fluent.py:560]: I think it's fine for this API to be stable; I don't imagine we'll want to rename or remove the `text` or `artifact_file` parameters

File: mlflow/tracking/client.py (hunk 2 above)
Comment (smurching) [reply mlflow/tracking/client.py:924]: Could you add a test for logging to such paths?

File: mlflow/tracking/client.py [hunk 8]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
+            tmp_path = os.path.join(tmp_dir, filename)
+            yield tmp_path
+            self.log_artifact(run_id, tmp_path, artifact_path)
+
+    def log_text(self, run_id, text, artifact_file):
+        """
+        Log a text as an artifact.
```
Comment (smurching) [inline mlflow/tracking/client.py:934]:
```suggestion
        Log text as an artifact.
```

File: mlflow/tracking/client.py [hunk 9]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
+
+        :param run_id: String ID of the run.
+        :param text: Text to log.
```
Comment (smurching) [inline mlflow/tracking/client.py:937]:
```suggestion
        :param text: String containing text to log.
```

File: mlflow/tracking/client.py (hunk 5 above)
Comment (harupy) [reply mlflow/tracking/client.py:927]: Found this comment stackoverflow:

https://stackoverflow.com/questions/13379742/right-way-to-clean-up-a-temporary-folder-in-python-class#comment114370145_19657028

File: mlflow/tracking/client.py [hunk 10]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
+
+        .. code-block:: python
+            :caption: Example
+
+            from mlflow.tracking import MlflowClient
+
+            client = MlflowClient()
+            run = client.create_run(experiment_id="0")
+
+            # Log a text in the artifact root
```
Comment (smurching) [inline mlflow/tracking/client.py:948]:
```suggestion
            # Log text to a file under the run's root artifact directory
```

File: mlflow/tracking/client.py (hunk 5 above)
Comment (smurching) [reply mlflow/tracking/client.py:927]: Oops, I missed

> On completion of the context or destruction of the temporary directory object the newly created temporary directory and all its contents are removed from the filesystem.

from the Python API doc...

File: mlflow/tracking/client.py [hunk 11]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
+            client.log_text(run.info.run_id, "text1", "file1.txt")
+
+            # Log a text in a subdirectory
```
Comment (smurching) [inline mlflow/tracking/client.py:951]:
```suggestion
            # Log text in a subdirectory of the run's root artifact directory
```

File: mlflow/tracking/client.py [hunk 12]
```diff
@@ -910,6 +914,50 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
...
+            client.log_text(run.info.run_id, "text2", "dir/file2.txt")
+
+            # Log an HTML text
```
Comment (smurching) [inline mlflow/tracking/client.py:954]:
```suggestion
            # Log HTML text
```

File: mlflow/tracking/client.py (hunk 2 above)
Comment (harupy) [reply mlflow/tracking/client.py:924]: done!

File: mlflow/tracking/fluent.py (hunk 4 above)
Comment (harupy) [reply mlflow/tracking/fluent.py:560]: Makes sense.

File: mlflow/tracking/client.py (hunk 3 above)
Comment (harupy) [reply mlflow/tracking/client.py]: Done in https://github.com/mlflow/mlflow/pull/3678/commits/b72d1c774d1dea7ebb0b227392b52ec4e586e169

File: mlflow/tracking/client.py [hunk 13]
```diff
@@ -910,6 +914,55 @@ def log_artifacts(self, run_id, local_dir, artifact_path=None):
         """
         self._tracking_client.log_artifacts(run_id, local_dir, artifact_path)
 
+    @contextlib.contextmanager
```
Comment (smurching) [inline mlflow/tracking/client.py:917]: BTW, what do you think about pulling these artifact logging utils into a private module, e.g. under `mlflow.artifacts._logging_utils` and actually defining `log_text` there as a static function, and then exposing it via the `MlflowClient` here via a single-line call to `mlflow.artifacts._logging_utils.log_text`? The advantage I think is that we can keep this file a bit more concise, especially as we add logging logic for more artifact types like images etc that may be a bit more complex

Comment (smurching) [reply mlflow/tracking/client.py:917]: Happy to have the refactor be in a follow-up PR or in this one, lmk what you think

**Review by smurching — APPROVED** [review]
LGTM, had one comment around moving artifact-logging utils to an internal module, but happy to address that in a follow up!

File: mlflow/tracking/client.py (hunk 13 above)
Comment (smurching) [reply mlflow/tracking/client.py:917]: To summarize offline discussion, this actually seems fine - it's hard to define `log_text` itself as a static function due to the dependency on the MlflowClient's `log_artifact`function.

From offline discussion:


> If anything in the future we could consider having utils for logging an in memory object to a pre-specified file path, & that could eventually become the pluggable interface for logging custom object types. You implement "log object type X to a specified file" and then MLflow knows how to use that method to log the object to a tempfile & call log_artifact on it. But I don't think we need to do that just yet

**Pull request merged**